SYCL/ESIMD/api/functional/ctors/common.hpp

```diff
@@ -29,54 +29,50 @@ using shared_allocator = sycl::usm_allocator<DataT, sycl::usm::alloc::shared>;
 template <typename DataT, int NumElems, typename TestCaseT>
-auto call_simd(sycl::queue &queue, const shared_vector<DataT> &ref_data) {
-
-  shared_vector<DataT> result{NumElems, shared_allocator<DataT>{queue}};
-
-  queue.submit([&](sycl::handler &cgh) {
-    const auto ref = ref_data.data();
-    auto out = result.data();
+bool verify_test_result(const std::vector<DataT> &ref_data,
+                        const shared_vector<DataT> &result_data,
+                        const std::string &data_type) {
+  bool passed{true};
+
+  for (size_t i = 0; i < result_data.size(); ++i) {
+    if (!are_bitwise_equal(ref_data[i], result_data[i])) {
+      passed = false;
+      log::fail<NumElems>(
+          "Simd by " + TestCaseT::get_description() +
+              " failed, retrieved: " + std::to_string(result_data[i]) +
+              ", expected: " + std::to_string(ref_data[i]) +
+              ", at index: " + std::to_string(i),
+          data_type);
+    }
+  }
 
-    cgh.single_task<Kernel<DataT, NumElems, TestCaseT>>(
-        [=]() SYCL_ESIMD_KERNEL {
-          sycl::ext::intel::experimental::esimd::simd<DataT, NumElems>
-              result_simd =
-                  TestCaseT::template call_simd_ctor<DataT, NumElems>(ref);
-          result_simd.copy_to(out);
-        });
-  });
-  return result;
+  return passed;
 }
 
 // The main test routine.
 // Using functor class to be able to iterate over the pre-defined data types.
 template <typename DataT, int NumElems, typename TestCaseT> struct test {
-  bool operator()(sycl::queue &queue, const std::string &data_type) {
+  template <typename VerifierT>
+  bool operator()(sycl::queue &queue, VerifierT verifier,
+                  const std::string &data_type) {
     bool passed{true};
-
-    std::vector<DataT> generated_data{generate_ref_data<DataT, NumElems>()};
-    shared_vector<DataT> ref_data{generated_data.begin(), generated_data.end(),
-                                  shared_allocator<DataT>{queue}};
-
-    const auto result_data =
-        call_simd<DataT, NumElems, TestCaseT>(queue, ref_data);
-
-    for (size_t it = 0; it < ref_data.size(); it++) {
-      if (!are_bitwise_equal(ref_data[it], result_data[it])) {
-        passed = false;
-        log::fail<NumElems>(
-            "Simd by " + TestCaseT::get_description() +
-                " failed, retrieved: " + std::to_string(result_data[it]) +
-                ", expected: " + std::to_string(ref_data[it]),
-            data_type);
+    std::vector<DataT> ref_data{generate_ref_data<DataT, NumElems>()};
+
+    // If current number of elements is equal to one, then run test with each
+    // one value from reference data.
+    // If current number of elements is greater than one, then run tests with
+    // whole reference data.
+    if constexpr (NumElems == 1) {
+      for (size_t i = 0; i < ref_data.size(); ++i) {
+        passed = verifier.template verify<DataT, NumElems, TestCaseT>(
+            queue, {ref_data[i]}, data_type);
```

Comment:
Do we need to call `std::vector` constructor only because of `verify_test_result` signature here? Or there is anything I'm missing from the first glance.

Reply:
No, previously we didn't created `std::vector` with single element in case `NumElems == 1`

Reply:
The `VerifierT` will be differ for the each test: for test on copy constructor, the `simd` required other simd object, but for the fill constructor requires pre-defined base value and step value
For many simd constructors requires different tests results verification logic

But for all tests, except test on `simd` default constructor I expect that we will have the same `run_test` function

Reply:
The previous solution was deleted, each test will have own `run_test` function, i've done it in this commit: 565f89ada9619b6ce9fc0cf6358a82a6fe707a73